File: src/Apiconvert.Core/Converters/RulesFormatter.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using Apiconvert.Core.Rules;

namespace Apiconvert.Core.Converters;

internal static class RulesFormatter
{
    internal static string FormatCanonical(object? rawRules, bool pretty = true)
    {
        var normalized = RulesNormalizer.NormalizeConversionRules(rawRules);
        var canonical = CanonicalizeRules(normalized);
        return canonical.ToJsonString(new JsonSerializerOptions(JsonDefaults.Options)
        {
            WriteIndented = pretty
        });
    }

    private static JsonObject CanonicalizeRules(ConversionRules rules)
    {
        var result = new JsonObject();

        if (rules.InputFormat != DataFormat.Json)
        {
            result["inputFormat"] = JsonValue.Create(rules.InputFormat.ToString().ToLowerInvariant());
        }

        if (rules.OutputFormat != DataFormat.Json)
        {
            result["outputFormat"] = JsonValue.Create(rules.OutputFormat.ToString().ToLowerInvariant());
        }

        var ruleNodes = new JsonArray();
        foreach (var rule in rules.Rules)
        {
            ruleNodes.Add(CanonicalizeRule(rule));
        }

        result["rules"] = ruleNodes;
        return result;
    }

    private static JsonObject CanonicalizeRule(RuleNode rule)
    {
        if (rule.Kind == "field")
        {
            var result = new JsonObject
            {
                ["kind"] = "field",
                ["outputPaths"] = ToJsonArray(rule.OutputPaths),
                ["source"] = CanonicalizeSource(rule.Source ?? new ValueSource())
            };

            if (!string.IsNullOrEmpty(rule.DefaultValue))
            {
                result["defaultValue"] = rule.DefaultValue;
            }

            return result;
        }

        if (rule.Kind == "array")
        {
            var result = new JsonObject
            {
                ["kind"] = "array",
                ["inputPath"] = rule.InputPath,
                ["outputPaths"] = ToJsonArray(rule.OutputPaths)
            };

            if (rule.CoerceSingle)
            {
                result["coerceSingle"] = true;
            }

            var itemRules = new JsonArray();
            foreach (var itemRule in rule.ItemRules)
            {
                itemRules.Add(CanonicalizeRule(itemRule));
            }

            result["itemRules"] = itemRules;
            return result;
        }

        var branch = new JsonObject
        {
            ["kind"] = "branch",
            ["expression"] = rule.Expression ?? string.Empty
        };

        var thenRules = new JsonArray();
        foreach (var thenRule in rule.Then)
        {
            thenRules.Add(CanonicalizeRule(thenRule));
        }

        branch["then"] = thenRules;

        if (rule.ElseIf.Count > 0)
        {
            var elseIf = new JsonArray();
            foreach (var elseIfRule in rule.ElseIf)
            {
                var elseIfNode = new JsonObject
                {
                    ["expression"] = elseIfRule.Expression ?? string.Empty
                };
                var elseIfThen = new JsonArray();
                foreach (var thenRule in elseIfRule.Then)
                {
                    elseIfThen.Add(CanonicalizeRule(thenRule));
                }

                elseIfNode["then"] = elseIfThen;
                elseIf.Add(elseIfNode);
            }

            branch["elseIf"] = elseIf;
        }

        if (rule.Else.Count > 0)
        {
            var elseRules = new JsonArray();
            foreach (var elseRule in rule.Else)
            {
                elseRules.Add(CanonicalizeRule(elseRule));
            }

            branch["else"] = elseRules;
        }

        return branch;
    }

    private static JsonObject CanonicalizeSource(ValueSource source)
    {
        var result = new JsonObject
        {
            ["type"] = source.Type
        };

        if (!string.IsNullOrEmpty(source.Path))
        {
            result["path"] = source.Path;
        }

        if (source.Paths.Count > 0)
        {
            result["paths"] = ToJsonArray(source.Paths);
        }

        if (source.Value is not null)
        {
            result["value"] = source.Value;
        }

        if (!string.IsNullOrEmpty(source.Expression))
        {
            result["expression"] = source.Expression;
        }

        if (source.TrueValue is not null)
        {
            result["trueValue"] = source.TrueValue;
        }

        if (source.FalseValue is not null)
        {
            result["falseValue"] = source.FalseValue;
        }

        if (source.TrueSource is not null)
        {
            result["trueSource"] = CanonicalizeSource(source.TrueSource);
        }

        if (source.FalseSource is not null)
        {
            result["falseSource"] = CanonicalizeSource(source.FalseSource);
        }

        if (source.ElseIf.Count > 0)
        {
            var elseIf = new JsonArray();
            foreach (var branch in source.ElseIf)
            {
                var branchNode = new JsonObject
                {
                    ["expression"] = branch.Expression ?? string.Empty
                };

                if (branch.Source is not null)
                {
                    branchNode["source"] = CanonicalizeSource(branch.Source);
                }

                if (branch.Value is not null)
                {
                    branchNode["value"] = branch.Value;
                }

                elseIf.Add(branchNode);
            }

            result["elseIf"] = elseIf;
        }

        if (source.ConditionOutput.HasValue)
        {
            result["conditionOutput"] = source.ConditionOutput.Value.ToString().ToLowerInvariant();
        }

        if (source.MergeMode.HasValue)
        {
            result["mergeMode"] = source.MergeMode.Value.ToString()[0].ToString().ToLowerInvariant() + source.MergeMode.Value.ToString()[1..];
        }

        if (source.Separator is not null)
        {
            result["separator"] = source.Separator;
        }

        if (source.TokenIndex.HasValue)
        {
            result["tokenIndex"] = source.TokenIndex.Value;
        }

        if (source.TrimAfterSplit.HasValue)
        {
            result["trimAfterSplit"] = source.TrimAfterSplit.Value;
        }

        if (source.Transform.HasValue)
        {
            result["transform"] = source.Transform.Value switch
            {
                TransformType.ToLowerCase => "toLowerCase",
                TransformType.ToUpperCase => "toUpperCase",
                TransformType.Number => "number",
                TransformType.Boolean => "boolean",
                TransformType.Concat => "concat",
                TransformType.Split => "split",
                _ => source.Transform.Value.ToString().ToLowerInvariant()
            };
        }

        if (!string.IsNullOrWhiteSpace(source.CustomTransform))
        {
            result["customTransform"] = source.CustomTransform;
        }

        if (source.Type == "merge" && !source.MergeMode.HasValue)
        {
            result["mergeMode"] = "concat";
        }

        if (source.Type == "condition" && !source.ConditionOutput.HasValue)
        {
            result["conditionOutput"] = "branch";
        }

        if (source.Type == "transform" && !source.Transform.HasValue && string.IsNullOrWhiteSpace(source.CustomTransform))
        {
            result["transform"] = "toLowerCase";
        }

        return result;
    }

    private static JsonArray ToJsonArray(IEnumerable<string> values)
    {
        var array = new JsonArray();
        foreach (var value in values)
        {
            array.Add(value);
        }

        return array;
    }
}

File: src/apiconvert-core/src/rules-formatter.ts
import { DataFormat, TransformType, type ConversionRules, type RuleNode, type ValueSource } from "./types";
import { normalizeConversionRules } from "./rules-normalizer";

export interface FormatConversionRulesOptions {
  pretty?: boolean;
}

export function formatConversionRules(rawRules: unknown, options: FormatConversionRulesOptions = {}): string {
  const rules = normalizeConversionRules(rawRules);
  const canonical = canonicalizeRules(rules);
  const pretty = options.pretty ?? true;
  return JSON.stringify(canonical, null, pretty ? 2 : 0);
}

function canonicalizeRules(rules: ConversionRules): Record<string, unknown> {
  const result: Record<string, unknown> = {};

  if ((rules.inputFormat ?? DataFormat.Json) !== DataFormat.Json) {
    result.inputFormat = rules.inputFormat;
  }

  if ((rules.outputFormat ?? DataFormat.Json) !== DataFormat.Json) {
    result.outputFormat = rules.outputFormat;
  }

  result.rules = (rules.rules ?? []).map(canonicalizeRule);
  return result;
}

function canonicalizeRule(rule: RuleNode): Record<string, unknown> {
  if (rule.kind === "field") {
    const result: Record<string, unknown> = {
      kind: "field",
      outputPaths: [...(rule.outputPaths ?? [])],
      source: canonicalizeSource(rule.source)
    };

    if (rule.defaultValue != null && rule.defaultValue.length > 0) {
      result.defaultValue = rule.defaultValue;
    }

    return result;
  }

  if (rule.kind === "array") {
    const result: Record<string, unknown> = {
      kind: "array",
      inputPath: rule.inputPath,
      outputPaths: [...(rule.outputPaths ?? [])],
      itemRules: (rule.itemRules ?? []).map(canonicalizeRule)
    };

    if (rule.coerceSingle) {
      result.coerceSingle = true;
    }

    return result;
  }

  const result: Record<string, unknown> = {
    kind: "branch",
    expression: rule.expression ?? "",
    then: (rule.then ?? []).map(canonicalizeRule)
  };

  if ((rule.elseIf ?? []).length > 0) {
    result.elseIf = (rule.elseIf ?? []).map((entry) => ({
      expression: entry.expression ?? "",
      then: (entry.then ?? []).map(canonicalizeRule)
    }));
  }

  if ((rule.else ?? []).length > 0) {
    result.else = (rule.else ?? []).map(canonicalizeRule);
  }

  return result;
}

function canonicalizeSource(source: ValueSource): Record<string, unknown> {
  const result: Record<string, unknown> = {
    type: source.type
  };

  if (source.path != null && source.path.length > 0) {
    result.path = source.path;
  }

  if (source.paths != null && source.paths.length > 0) {
    result.paths = [...source.paths];
  }

  if (source.value != null) {
    result.value = source.value;
  }

  if (source.expression != null && source.expression.length > 0) {
    result.expression = source.expression;
  }

  if (source.trueValue != null) {
    result.trueValue = source.trueValue;
  }

  if (source.falseValue != null) {
    result.falseValue = source.falseValue;
  }

  if (source.trueSource != null) {
    result.trueSource = canonicalizeSource(source.trueSource);
  }

  if (source.falseSource != null) {
    result.falseSource = canonicalizeSource(source.falseSource);
  }

  if (source.elseIf != null && source.elseIf.length > 0) {
    result.elseIf = source.elseIf.map((entry) => {
      const branch: Record<string, unknown> = {
        expression: entry.expression ?? ""
      };
      if (entry.source != null) {
        branch.source = canonicalizeSource(entry.source);
      }
      if (entry.value != null) {
        branch.value = entry.value;
      }
      return branch;
    });
  }

  if (source.conditionOutput != null) {
    result.conditionOutput = source.conditionOutput;
  }

  if (source.mergeMode != null) {
    result.mergeMode = source.mergeMode;
  }

  if (source.separator != null) {
    result.separator = source.separator;
  }

  if (source.tokenIndex != null) {
    result.tokenIndex = source.tokenIndex;
  }

  if (source.trimAfterSplit != null) {
    result.trimAfterSplit = source.trimAfterSplit;
  }

  if (source.transform != null) {
    result.transform = source.transform;
  }

  if (source.customTransform != null && source.customTransform.length > 0) {
    result.customTransform = source.customTransform;
  }

  if (source.type === "merge" && source.mergeMode == null) {
    result.mergeMode = "concat";
  }

  if (source.type === "condition" && source.conditionOutput == null) {
    result.conditionOutput = "branch";
  }

  if (source.type === "transform" && source.transform == null && source.customTransform == null) {
    result.transform = TransformType.ToLowerCase;
  }

  return result;
}
